# Assignment-1 (KNN Classifier)--PartB
## Breast cancer wisconsin dataset
In this assignment, we will build a KNN classifier that takes an features as as input and outputs a label 0 or 1.

The breast cancer dataset contains 569 samples with 30 real, positive features (including cancer mass attributes like mean radius, mean texture, mean perimeter, et cetera). Of the samples, 212 are labeled “malignant” and 357 are labeled “benign”. 
You can find more details in: https://scikit-learn.org/stable/datasets/index.html#breast-cancer-dataset

In [45]:
from sklearn.datasets import load_breast_cancer
import numpy as np
import time
## Load the training set
data = load_breast_cancer()
X = data.data
y = data.target


In [19]:
## print some statistics on the dataset
### TODO YOUR CODE ###

print("Total number of samples: ", len(X))

# print the number of features per sample
print("number of features", np.shape(X)[1])


# Total number of classes
print("the number of classes: ", len(np.unique(y)))



# print the number of samples in each class
# print(y)
positive = sum(y)
negative = len(y) - positive
print("the number of samples in the first class: ", positive)
print("the number of samples in the second class: ", negative)

Total number of samples:  569
number of features 30
the number of classes:  2
the number of samples in the first class:  357
the number of samples in the second class:  212


## Splitting the Train data to Train and Validate Sets

In [20]:
from sklearn.model_selection import train_test_split
### TODO YOUR CODE ###

# Split the data (i.e. features and target) into 70% train, 15% validate, and 15% test; Use Random Seed 777

X_train, X_temp, y_train, y_temp = train_test_split(X, y, test_size=0.3, random_state=777)

X_val, X_test, y_val, y_test = train_test_split(X_temp, y_temp, test_size=0.5, random_state=777)


## Nearest neighbor classification with L2 distance

To compute nearest neighbors in our data set, we need to first be able to compute distances between data points. A natural distance function is _Euclidean distance_: for two vectors $x, y \in \mathbb{R}^d$, their Euclidean distance is defined as 
$$\|x - y\| = \sqrt{\sum_{i=1}^d (x_i - y_i)^2}.$$
Often we omit the square root, and simply compute _squared Euclidean distance_:
$$\|x - y\|^2 = \sum_{i=1}^d (x_i - y_i)^2.$$
For the purposes of nearest neighbor computations, the two are equivalent: for three vectors $x, y, z \in \mathbb{R}^d$, we have $\|x - y\| \leq \|x - z\|$ if and only if $\|x - y\|^2 \leq \|x - z\|^2$.

## 1. Nearest neighbor classification with L2 distance

Write a function, **NN_L2**, which takes as input the training data (`trainx` and `trainy`) and the test points (`evalx`) and predicts labels for these test points using 1-NN classification. These labels should be returned in a `numpy` array with one entry per test point. For **NN_L2**, the L2 norm should be used as the distance metric.

In [21]:
def NN_L2(trainx, trainy, evalx):
    # inputs: trainx, trainy, testx <-- as defined above
    # output: an np.array of the predicted values for testy 
    ### START CODE HERE ###
    predictions = []
    
    for eval_point in evalx:
        distances = np.sqrt(np.sum((trainx - eval_point) ** 2, axis=1))
        nearest_neighbor_index = np.argmin(distances)
        predicted_label = trainy[nearest_neighbor_index]
        
        predictions.append(predicted_label)
    
    return np.array(predictions)
    
    ### END CODE HERE ###

## 2. K-Nearest neighbor classification with L2 distance

Write a function, **KNN_L2**, which takes as input the training data (`trainx` and `trainy`), the test points (`evalx`), and the value of **K** (integer) and predicts labels for these test points using K-NN classification. These labels should be returned in a `numpy` array with one entry per test point.

In [22]:
def KNN_L2(trainx, trainy, evalx, K):
    
    ### START CODE HERE ###

    predictions = []
    
    for eval_point in evalx:
        distances = np.sqrt(np.sum((trainx - eval_point) ** 2, axis=1))
        
        nearest_neighbor_indices = np.argsort(distances)[:k]
        
        nearest_neighbor_labels = trainy[nearest_neighbor_indices]
        
        predicted_label = np.argmax(np.bincount(nearest_neighbor_labels))
        
        predictions.append(predicted_label)
    
    return np.array(predictions)
    ### END CODE HERE ###

## 3. Nearest neighbor classification with L1 distance

We now compute nearest neighbors using the L1 distance (sometimes called *Manhattan Distance*).

Write a function, **NN_L1**, which again takes as input the arrays `trainx`, `trainy`, and `evalx`, and predicts labels for the test points using 1-nearest neighbor classification. For **NN_L1**, the L1 distance metric should be used. As before, the predicted labels should be returned in a `numpy` array with one entry per test point.

Notice that **NN_L1** and **NN_L2** may well produce different predictions on the test set.

In [23]:
def NN_L1(trainx, trainy, evalx):
    # inputs: trainx, trainy, testx <-- as defined above
    # output: an np.array of the predicted values for testy 
    
    ### START CODE HERE ###
    predictions = []
    
    for eval_point in evalx:
          distances = np.sum(np.abs(trainx - eval_point), axis=1)
        
        nearest_neighbor_index = np.argmin(distances)
        
        predicted_label = trainy[nearest_neighbor_index]
        
        predictions.append(predicted_label)
    
    return np.array(predictions)
    ### END CODE HERE ###

## 4. K-Nearest neighbor classification with L1 distance

Write a function, **KNN_L1**, which takes as input the training data (`trainx` and `trainy`), the test points (`evalx`), and the value of **K** (integer) and predicts labels for these test points using K-NN classification and L1 distance metric. These labels should be returned in a `numpy` array with one entry per test point.

In [24]:
def KNN_L1(trainx, trainy, evalx, K):
    
    ### START CODE HERE ###
    predictions = []
    
    for eval_point in evalx:
        distances = np.sum(np.abs(trainx - eval_point), axis=1)
        
        nearest_neighbor_indices = np.argsort(distances)[:K]
        
        nearest_neighbor_labels = trainy[nearest_neighbor_indices]
        
        predicted_label = np.argmax(np.bincount(nearest_neighbor_labels))
        
        predictions.append(predicted_label)
    
    return np.array(predictions)
    ### END CODE HERE ###

## 5. K-Nearest neighbor classifier

Write a function, **KNN**, which takes as input the training data (`trainx` and `trainy`), the test points (`evalx`), the value of **K** (integer), and a parameter for deciding the distance metric to be used (for example 1 for L1 and 2 for L2) and predicts labels for these test points using KNN classification. These labels should be returned in a `numpy` array with one entry per test point.

In [25]:
def KNN(trainx, trainy, evalx, K, dist_metric=2):
    
    ### START CODE HERE ###
    if dist_metric == 1:
        return KNN_L1(trainx, trainy, evalx, K)
    elif dist_metric == 2:
        return KNN_L2(trainx, trainy, evalx, K)
    else:
        print("worng value chose either 1, or 2")

    ### END CODE HERE ###

## 6. Putting it all together

Write code that allows you to select the hyper-parameters (distance measure and the value of K) by calling the KNN classifier with different values of K and either L1 or L2 distance measure. Make sure that you set the hyper-parameters using the validation set and not the test set. You need to systemtically try different values for K in conjunction with a distance measure and tabulate the results (you can do that be craeting a seperate cell and documenting in that cell) and note down the best hyper-parameter settings.

In [28]:
### START CODE HERE ###
def accuracy(y_pred, valy):
    if len(y_pred) != len(valy):
        print("Error, different Sizes!")
        return None
    correct = np.sum(y_pred == valy)
    accuracy = correct / len(y_pred)
    return accuracy
L1acc = []
L2acc = []
Ks = [k for k in range(1, 26) if k % 2 != 0]
for k in Ks: 
    y_predL1 = KNN(X_train, y_train, X_val,k , dist_metric=1)
    y_predL2 = KNN(X_train, y_train, X_val, k)
    L1acc.append(accuracy(y_predL1, y_val))
    L2acc.append(accuracy(y_predL2, y_val))
    
best_acc_L1 = max(L1acc)
best_acc_L2 = max(L2acc)
best_K_L1 = Ks[L1acc.index(best_acc_L1)]
best_K_L2 = Ks[L2acc.index(best_acc_L2)]
# X_train, X_temp, y_train, y_temp 

# X_val, X_test, y_val, y_test 


### END CODE HERE ###

In [48]:
print(dict(zip(Ks, L1acc)))
print(dict(zip(Ks, L2acc)))

print("the best accuracy along withe the K for the L1 metric", best_acc_L1,"and",  best_K_L1)
print("the best accuracy along withe the K for the L2 metric", best_acc_L2,"and",  best_K_L2)
print("we will be using L1 with K = 5")

{1: 0.9176470588235294, 3: 0.9176470588235294, 5: 0.9529411764705882, 7: 0.9411764705882353, 9: 0.9176470588235294, 11: 0.9176470588235294, 13: 0.9176470588235294, 15: 0.9176470588235294, 17: 0.8941176470588236, 19: 0.9058823529411765, 21: 0.9058823529411765, 23: 0.8941176470588236, 25: 0.8941176470588236}
{1: 0.8941176470588236, 3: 0.9411764705882353, 5: 0.9294117647058824, 7: 0.9294117647058824, 9: 0.9176470588235294, 11: 0.9176470588235294, 13: 0.9058823529411765, 15: 0.9176470588235294, 17: 0.9058823529411765, 19: 0.8823529411764706, 21: 0.8823529411764706, 23: 0.8941176470588236, 25: 0.8941176470588236}
the best accuracy along withe the K for the L1 metric 0.9529411764705882 and 5
the best accuracy along withe the K for the L2 metric 0.9411764705882353 and 3
we will be using L1 with K = 5


# 7. Test errors and the confusion matrix

Once the hyper-parameters have been selected, we now would like to perform a final evaluation on the test set and record the error rates. Also, Write a function, **confusion**, which takes as input the true labels for the test set (that is, `testy`) as well as the predicted labels and returns the confusion matrix. The confusion matrix should be a `np.array`.

**Note:** Record the cpu time it takes to perform the evaluation on the test set using functions like **time.time()**.

In [42]:
def confusion(testy, testy_fit):
    # inputs: the correct labels, the fitted KNN labels 
    # output: a 10x10 np.array representing the confusion matrix as above
    
    ### START CODE HERE ###
    n_classes = len(np.unique(testy))  # Get the number of unique classes
    confusion_matrix = np.zeros((n_classes, n_classes), dtype=int)

    for true_label, predicted_label in zip(testy, testy_fit):
        confusion_matrix[true_label][predicted_label] += 1

    return confusion_matrix
    ### END CODE HERE ###

In [50]:
# Code for performing the final evaluation on the test set and generating the confuson matrix.
### START CODE HERE ###
# X_val, X_test, y_val, y_test 
startTime = time.process_time()
y_pred =KNN(X_train, y_train, X_test, 5, dist_metric=1)
endTime = time.process_time()
print("accuracy on test: ", accuracy(y_pred, y_test))
allTime = endTime - startTime
arr =  confusion(y_test, y_pred)
print(arr)
print("the test time ")
print(allTime)
### END CODE HERE ###

accuracy on test:  0.9302325581395349
[[23  6]
 [ 0 57]]
the test time 
0.0


## 8. Preparing the assignment report

You need to record your answers for the following questions:
1. What is the error rate on the validation set for NN_L2?
1: 0.8941176470588236
2. What is the best error rate on the validation set for KNN_L2?
3: 0.9411764705882353
3. What is the error rate on the validation set for NN_L1?
1: 0.9176470588235294
4. What is the best error rate on the validation set for KNN_L1?
5: 0.9529411764705882
5. What is the error rate on the test set?
0.9302325581395349
7. Do you need to normalize data, in general, when using KNN?
ans: depeneds, sometimes so that not a single feature dominate the testing phase, making all the other features irreleavent
8. Do you need to normalize data when using KNN for the given problem? Explain why? 
appearently we don't since the performance is alright, not too bad, not too great

## 9. Extra Stuff

You are invited to try some more ideas as extra work like:
1. Implementing weighted KNN where the vote of a neighbour is scaled down based on its distance from the test point.
2. Implement L_infinity distance measure and use it for classification.

## Points to remember

You need to keep in mind the following points:
1. Use numpy arrays and numpy libraries for efficient computations. 
2. Vectorize the code wherever possible instead of using explicit loops. This will significantly speed-up your code.